File: note/이진탐색.js
// 데이터가 정렬되어 있는 배열에서 특정한 값을 찾아내는 알고리즘
// 배열의 중간에 있는 임의의 값을 선택하고 찾고자하는 값 x와 비교
// 1. x가 중간값보다 작으면 중간값을 기준으로 좌측의 데이터를 대상으로 함
// 2. x가 중간값보다 크면 중간값을 기준으로 우측의 데이터를 대상으로 함
// 찾을 때까지 반복
// 이진탐색(O(logN))은 단순한 배열 순회(O(N))보다 시간복잡도 상으로 크게 이점을 가짐

// 구현
const binarySearch = function(arr, target) {
  let start = 0;
  let end = arr.length-1;
  let mid

  while(start<=end) {
    mid = parseInt((start+end)/2);
    if(target === arr[mid]) {
      return mid;
    } else {
      if(target < arr[mid]) { // 좌측
        end = mid-1;
      } else {  // 우측
        start = mid+1;
      }
    }
  }
  return -1;
}

let output = binarySearch([1,2,3,4,5,6,7,8,9],2);
// 인덱스 값 반환
console.log(output);

File: note/이진 탐색.py
# 정렬된 배열 또는 리스트에서 특정 값을 빠르게 찾는 효율적인 알고리즘
# 큰 데이터 셋에서 매우 빠르게 동작
# 시간 복잡도
# O(log n)

# 작동 원리
# 1. 중간 원소 선택
# 2. 비교 : 중간 원소와 찾고자 하는 값을 비교
# 중간 원소 = 값 : 검색 종료, 인덱스 반환
# 중간 원소 < 값 : 중간 원소의 오른쪽 절반으로 줄임
# 중간 원소 > 값 : 중간 원소의 왼쪽 절반으로 줄임
# 3. 반복

# 구현
# 1. 반복적 방법
def binary_search(arr, target):
    left, right = 0, len(arr) - 1

    while left <= right:
        mid = (left + right) // 2
        if arr[mid] == target:
            return mid
        elif arr[mid] < target:
            left = mid + 1
        else:
            right = mid - 1

    return -1  # 찾고자 하는 값이 배열에 없는 경우

# 예제 사용
arr = [2, 3, 4, 10, 40]
target = 10
result = binary_search(arr, target)
print(f"Element is present at index {result}" if result != -1 else "Element is not present in array")

# 2. 재귀적 방법

def binary_search_recursive(arr, target, left, right):
    if left <= right:
        mid = (left + right) // 2
        if arr[mid] == target:
            return mid
        elif arr[mid] < target:
            return binary_search_recursive(arr, target, mid + 1, right)
        else:
            return binary_search_recursive(arr, target, left, mid - 1)
    else:
        return -1  # 찾고자 하는 값이 배열에 없는 경우

# 예제 사용
arr = [2, 3, 4, 10, 40]
target = 10
result = binary_search_recursive(arr, target, 0, len(arr) - 1)
print(f"Element is present at index {result}" if result != -1 else "Element is not present in array")
